controller/appcontroller.go

```diff
@@ -2316,6 +2321,34 @@ func (ctrl *ApplicationController) newApplicationInformerAndLister() (cache.Shar
 	if err != nil {
 		return nil, nil
 	}
+
+	// Periodically updates ObservedAt time in Application health status to indicate freshness
+	_, err = informer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(old, new interface{}) {
+				if !ctrl.canProcessApp(new) {
+					return
+				}
+				oldApp, oldOK := old.(*appv1.Application)
+				newApp, newOK := new.(*appv1.Application)
+				if oldOK && newOK {
+					// To prevent a status update hot loop, enqueue only events generated by resyncs
+					// and ignore those triggered by application updates, which are handled by another event handler.
+					if oldApp.ResourceVersion != newApp.ResourceVersion {
+						return
+					}
+
+					// We are interested only in refreshing the status, not performing a resource-level comparison.
+					ctrl.requestAppRefresh(newApp.QualifiedName(), ComparisonWithNothing.Pointer(), nil)
```

Comment:
Is there a way to perhaps only do the refresh while a sync is taking place for the app? As it stands now, this will refresh every 30 seconds throughout the app's lifecycle.

Reply:
> Are you suggesting updating the `.status.health.observedAt` field only when the app is syncing? 

Yes exactly 😄 

> However, I’m not sure how expensive that operation is.

For a single app it's likely negligible, but @crenshaw-dev raised the concern that if there were say 10k apps, every 30 secs there would be 10k requests hammering the K8s API server (updating the `.status.health.observedAt` field).

Reply:
Yeah, that's a fair concern.

> The hope is to sidestep that concern by only performing this refresh during app sync.

I believe we can achieve this without significant changes to the current implementation by removing the new `AddEventHandlerWithResyncPeriod` event handler. However, without a periodic refresh, I am not sure how much value this new field provides. Existing fields already provide the same information that `.status.health.observedAt` would, if updated only during syncs.

Another approach could be to enable updates to this field during app sync by default and put the periodic update behind a flag.

Reply:
> without a periodic refresh, I am not sure how much value this new field provides

My understanding of the 30s refresh is that it's just a "safety net." My assumption would be that the health is constantly being refreshed based changes in the cluster cache and that the 30s refresh would only catch updates that were missed, which I think we would expect to be a rare event.

Reply:
Nevermind, I read [above](https://github.com/argoproj/argo-cd/pull/18660#discussion_r1657781119) that the intent is to bump the `observedAt` timestamp to prove that the health has been evaluated more recently than the last sync operation.

I worry about accruing tech debt (with a feature behind a flag) or incurring a performance penalty (with a feature without a flag) to provide a service answering the question "has this app been evaluated to be healthy since the app was synced?" There are already a few ways already to do this:

1) Wait a reasonable amount of time after the app has synced to observe the health (makes it very likely the health was evaluated)
2) Request an app refresh and then observe the health (guarantees the health was evaluated)
3) Add a post-sync resource hook to the app (ties the sync success/failure to resource health)

Reply:
Sure... I think it's _nice_ for Argo CD to produce a fairly-recent timestamp for the most recent observed health. I just wanna make sure it's worth the performance and maintenance costs as compared to other options. Do we have some sense for a typical `comparison_with_nothing_ms` distribution on a fairly-busy Argo instance? Is there reason to believe we might need the heartbeat time to be configurable or to include jitter? I'm in favor of spending CPU cycles on useful stuff, I just think we should first have a clear sense for how many CPU cycles we're spending.

Reply:
Eh, looks like that checkpoint was only added in 2.13, so I don't have any stats from Intuit.

I'm not super worried about it tbh. I'd just recommend trying it out on some relatively busy Argo instance and watching the effect on both the k8s API and Argo controller performance.